### PR TITLE
Zero-initialize aggregate states with destructors immediately after allocating

### LINKED ISSUE #10022: duckdb segmentation fault (crash) with approx_count_distinct function when query throws out of memory error
### What happens?

when you run a query with approx_count_distinct function on a duck db database and out of memory condition comes then instead of throwing out of memory error, duck db crashes with segmentation fault

### To Reproduce

1. Launch terminal/command prompt/shell and go the directory where duckdb exe is present
2. Attach to the duckdb database using below command. You can find lineorder_flat.db file in the attachment (attachment contains .txt extension, so please remove .txt extension): [lineorder_flat.db.txt](https://github.com/duckdb/duckdb/files/13817506/lineorder_flat.db.txt)
[lineorder_flat.db.txt](https://github.com/duckdb/duckdb/files/13817506/lineorder_flat.db.txt)

This file contains empty table
`./duckdb lineorder_flat.db`

3. Insert data to the table using below command and [csv file, de-normalised SSB line order data](https://community-shared-data-1308875761.cos.ap-beijing.myqcloud.com/lineorder_flat.tar.bz2)

`INSERT INTO lineorder_flat SELECT * FROM read_csv('lineorder_flat.tbl', AUTO_DETECT=TRUE);`

4. Set max memory to the database using following SQL. In this case I set max memory as 400 MB to reproduce the crash but this crash can come with any memory configuration provided that the query should generate out of memory condition.
`set max_memory="400MB"`

5. Run the query that causes out of memory error. In this case I ran below query

`Select lo_orderkey, lo_linenumber, lo_custkey, lo_partkey, lo_suppkey, lo_orderdate, lo_commitdate, c_nation, c_name, c_address, c_city, c_nation, c_region, c_phone, c_mktsegment, s_name, s_address, s_city, s_nation, s_region, s_phone, p_name, p_mfgr, p_category, p_brand, p_color, p_type, p_size, p_container, sum(lo_tax), sum(lo_discount), sum(lo_extendedprice), approx_count_distinct(c_name) from lineorder group by all;`

6. Please note that if instead of "approx_count_distinct" if I use "count" function, query will throw out of memory error. 
7. Also note that this crash may not happen every time but it is reproducing very frequently. In rare case, it throws out of memory error
8. I have also attached duckdb crash dump file that I obtained from core dump. 
[crashdump.txt](https://github.com/duckdb/duckdb/files/13700120/crashdump.txt)

### OS:

MaCOS

### DuckDB Version:

Latest

### DuckDB Client:

command line

### Full Name:

Mozz

### Affiliation:

Self

### Have you tried this on the latest `main` branch?

I have tested with a main build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

### PR DESCRIPTION
**Pull request by lnkuiper**
Fixes #10022

### CODE REVIEW

File: src/common/types/row/tuple_data_layout.cpp [hunk 1]
```diff
@@ -110,10 +111,11 @@ void TupleDataLayout::Initialize(vector<LogicalType> types_p, Aggregates aggrega
 #endif
 
 	has_destructor = false;
-	for (auto &aggr : GetAggregates()) {
+	for (idx_t aggr_idx = 0; aggr_idx < aggregates.size(); aggr_idx++) {
+		const auto &aggr = aggregates[aggr_idx];
 		if (aggr.function.destructor) {
 			has_destructor = true;
-			break;
+			aggr_destructor_idxs.push_back(aggr_idx);
```
Comment (Mytherin) [inline src/common/types/row/tuple_data_layout.cpp:118]: Minor nit: maybe `has_destructor` can be removed now and replaced with `!aggr_destructor_idxs.empty()`?

Comment (lnkuiper) [reply src/common/types/row/tuple_data_layout.cpp:114]: Good nit, thanks. Fixed

File: src/include/duckdb/common/types/row/tuple_data_layout.hpp [hunk 2]
```diff
@@ -89,7 +92,11 @@ class TupleDataLayout {
 	}
 	//! Returns whether any of the aggregates have a destructor
 	inline bool HasDestructor() const {
-		return has_destructor;
+		return !aggr_destructor_idxs.empty();
+	}
+	//! Returns the indices of the aggregates that have destructors
+	inline const vector<idx_t> GetAggregateDestructorIndices() const {
```
Comment (Mytherin) [inline src/include/duckdb/common/types/row/tuple_data_layout.hpp:98]: nit: Maybe `const vector<idx_t>&`

**Comment by hawkfish** [comment]
Have we checked that all aggregate states have POD? For example, `QuantileState` is full of `vector`s and smart pointers.

**Pull request merged**